core/src/main/java/org/hibernate/ogm/dialect/eventstate/impl/EventContextManager.java

```diff
@@ -72,7 +72,9 @@ void onEventFinished() {
 
 		SessionImplementor session = (SessionImplementor) states.get( SessionImplementor.class );
 
-		for ( Entry<Class<?>, Object> state : states.entrySet() ) {
+		// we need to make a copy of the map as an OperationCollector entry might be added during
```

Comment:
In which case does this exactly happen? I'm no fan of this right now, because - as you say below - the onFinish() will not be called for the newly added state. Seems like a potential cause for trouble.